lib/common/output_text.c

```diff
@@ -0,0 +1,200 @@
+
+    va_start(ap, format);
+
+    /* Informational output does not get indented, to separate it from other
+     * potentially indented list output.
+     */
+    len = vfprintf(out->dest, format, ap);
+    CRM_ASSERT(len > 0);
+    va_end(ap);
+}
+
+static void
+text_output_xml(pcmk__output_t *out, const char *name, const char *buf) {
+    text_private_t *priv = out->priv;
+
+    CRM_ASSERT(priv != NULL);
+    pcmk__indented_printf(out, "%s", buf);
+}
+
+static void
+text_begin_list(pcmk__output_t *out, const char *header, const char *singular_noun, const char *plural_noun) {
+    text_private_t *priv = out->priv;
+    text_list_data_t *new_list = NULL;
+
+    CRM_ASSERT(priv != NULL);
+
+    pcmk__indented_printf(out, "%s:\n", header);
+
+    new_list = calloc(1, sizeof(text_list_data_t));
+    new_list->len = 0;
+    new_list->singular_noun = strdup(singular_noun);
+    new_list->plural_noun = strdup(plural_noun);
+
+    g_queue_push_tail(priv->parent_q, new_list);
+}
+
+static void
+text_list_item(pcmk__output_t *out, const char *id, const char *content) {
+    text_private_t *priv = out->priv;
+
+    CRM_ASSERT(priv != NULL);
+
+    if (id != NULL) {
+        pcmk__indented_printf(out, "* %s: %s\n", id, content);
+    } else {
+        pcmk__indented_printf(out, "* %s\n", content);
+    }
+
+    ((text_list_data_t *) g_queue_peek_tail(priv->parent_q))->len++;
+}
+
+static void
+text_end_list(pcmk__output_t *out) {
+    text_private_t *priv = out->priv;
+    text_list_data_t *node = NULL;
+
+    CRM_ASSERT(priv != NULL);
+    node = g_queue_pop_tail(priv->parent_q);
+
+    if (node->singular_noun != NULL && node->plural_noun != NULL) {
+        if (node->len == 1) {
+            pcmk__indented_printf(out, "%d %s found\n", node->len, node->singular_noun);
+        } else {
+            pcmk__indented_printf(out, "%d %s found\n", node->len, node->plural_noun);
+        }
+    }
+
+    free(node);
+}
+
+pcmk__output_t *
+pcmk__mk_text_output(char **argv) {
+    pcmk__output_t *retval = calloc(1, sizeof(pcmk__output_t));
+
+    if (retval == NULL) {
+        return NULL;
+    }
+
+    retval->request = g_strjoinv(" ", argv);
+    retval->supports_quiet = true;
+
+    retval->init = text_init;
+    retval->free_priv = text_free_priv;
+    retval->finish = text_finish;
+    retval->reset = text_reset;
+
+    retval->register_message = pcmk__register_message;
+    retval->message = pcmk__call_message;
+
+    retval->command_output = text_command_output;
+    retval->info = text_info;
+    retval->output_xml = text_output_xml;
+
+    retval->begin_list = text_begin_list;
+    retval->list_item = text_list_item;
+    retval->end_list = text_end_list;
+
+    return retval;
+}
+
+G_GNUC_PRINTF(2, 3)
+void
+pcmk__indented_printf(pcmk__output_t *out, const char *format, ...) {
+    va_list ap;
+    int len = 0;
+    text_private_t *priv = out->priv;
+
+    CRM_ASSERT(priv != NULL);
+
+    for (int i = 0; i < g_queue_get_length(priv->parent_q); i++) {
```

Comment:
I believe g_queue_get_length() will be called for each iteration here, not a big deal since this will rarely get beyond 3 or maybe 4, but best practice would be to call it once before the loop

Reply:
I was thinking the compiler would be smart enough to optimize that, but it's easy enough to change.